src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,542 @@
+    unitsUserSchema = hGrp->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feet
+    hGrp->Attach(this);
+}
+
+GridExtensionP::~GridExtensionP()
+{
+    Gui::coinRemoveAllChildren(GridRoot);
+    GridRoot->unref();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    hGrp->Detach(this);
+}
+
+double GridExtensionP::getGridSize() const
+{
+    return computedGridValue;
+}
+
+void GridExtensionP::getClosestGridPoint(double &x, double &y) const
+{
+    auto closestdim = [](double &dim, double gridValue) {
+        dim = dim / gridValue;
+        dim = dim < 0.0 ? ceil(dim - 0.5) : floor(dim + 0.5);
+        dim *= gridValue;
+    };
+
+    closestdim(x, computedGridValue);
+    closestdim(y, computedGridValue);
+
+    //Base::Console().Log("gridvalue=%f, (x,y)=(%f,%f)", computedGridValue, x, y);
+}
+
+bool GridExtensionP::checkCameraZoomChange(const Gui::View3DInventorViewer* viewer)
+{
+    float newCamMaxDimension = viewer->getMaxDimension();
+    if (fabs(newCamMaxDimension - camMaxDimension) > 0) { //ie if user zoomed.
+        camMaxDimension = newCamMaxDimension;
+        return true;
+    }
+
+    return false;
+}
+
+bool GridExtensionP::checkCameraTranslationChange(const Gui::View3DInventorViewer* viewer)
+{
+    //Then we check if user moved by more than 10% of camera dimension (must be after updating camera dimension).
+    SbVec3f newCamCenterOnSketch = viewer->getCenterPointOnFocalPlane();
+
+    if ((camCenterOnSketch - newCamCenterOnSketch).length() > 0.1 * camMaxDimension) {
+        camCenterOnSketch = newCamCenterOnSketch;
+        return true;
+    }
+
+    return false;
+}
+
+void GridExtensionP::computeGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    if (!vp->GridAuto.getValue()) {
+        computedGridValue = vp->GridSize.getValue();
+        return;
+    }
+
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0) {
+        computedGridValue = vp->GridSize.getValue();
+        return;
+    }
+
+    int numberOfLines = static_cast<int>(std::max(pixelWidth, pixelHeight)) / GridSizePixelThreshold;
+
+    double unitMultiplier = (unitsUserSchema == 2 || unitsUserSchema == 3) ? 25.4 : (unitsUserSchema == 5 || unitsUserSchema == 7) ? 304.8 : 1;
+
+    // If number of subdivision is 1, grid auto spacing can't work as it uses it as a factor
+    // In such case, we apply a default factor of 10
+    auto safeGridNumberSubdivision = GridNumberSubdivision <= 1 ? 10 : GridNumberSubdivision;
+
+    computedGridValue = vp->GridSize.getValue() * safeGridNumberSubdivision * unitMultiplier * pow(safeGridNumberSubdivision, floor(log(camMaxDimension / unitMultiplier / numberOfLines / vp->GridSize.getValue()) / log(safeGridNumberSubdivision)));
```

Comment:
```suggestion
    // If number of subdivision is 1, grid auto spacing can't work as it uses it as a factor
    // In such case, we apply a default factor of 10
    auto safeGridNumberSubdivision = GridNumberSubdivision <= 1 ? 10 : GridNumberSubdivision;

    computedGridValue = vp->GridSize.getValue() * safeGridNumberSubdivision * pow(safeGridNumberSubdivision, floor(log(camMaxDimension / numberOfLines / vp->GridSize.getValue()) / log(safeGridNumberSubdivision)));
```
I propose this change because :
* I can't give a mathematical sense to current formula
* This doesn't cause any regression to current behavior
* It actually improves things. With current formula, for example for a grid size of 1 inch and with 'Auto scale', the drawn grid has very weird spacing, never aligned with "normal" 1 inch grid. After change, the drawn grid is aligned with 1-inch grid.

Reply:
> OK, I can reproduce. As I thought, there as been a glitch where the GridSize value came to 0. It can be reproduced by using the Parameter Editor, setting the GridSize to 0, then creating a sketch and enabling the grid.
> 
> What shall we do in such a situation ? Just not draw the grid ? Or draw with default value ? Shall we warn user ?

I think the simplest solution is to warn and return. 

However I am wondering which one would be a good low enough value to compare . The widget does not allow a value smaller than 1 um. I wonder if this is right too. For OCCT with have Precision::Confusion(). But I am not sure we have something for pure visualisation using coin...  perhaps

```std::numeric_limits<float>::min();``` ?

Reply:
I think everything small enough is fine. We can change later. Everything is better than a hard crash.

Reply:
Oh! We already have limited computedGridValue by hardcoded values in computeGridSize (currently 1 um and 10000 km).

The behaviour we have in code is to default to that hardcoded value if outside range.

Are these values sound?

Reply:
Let's say they are. Not a big deal to change later.
PS: just notice this discussion is probably not happening in the right place. The original comment isn't really linked. 😋

Reply:
Sorry for that.

Reply:
@PaddleStroke 

Three posts above (I commented on the wrong thread), @0penBrain comments on the unitMultiplier.

He cannot make sense of it. However, I think you must have put it there for a good reason.

Would you please elaborate why you added it?